[vcpkg] Improve make builds

Neumann-A opened the pull request:
Trying to correctly fix make builds for msvc

related to #9966 (main logic to make msvc builds work are taken from there)

Neumann-A committed:
update all 16 configure_make ports

Neumann-A committed:
add make wrappers for msvc

Neumann-A committed:
improve make builds

JackBoosY commented:
Thanks @Neumann-A !

Neumann-A committed:
fix relativ path errors on linux (and osx?)

Neumann-A committed:
revisit all 16 portfiles again

Neumann-A commented:
hmm some of those build failures I am unable to reproduce locally 

This one is also kind of strange and only appears in some ports: 
```
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:58 (file):
  file failed to open for reading (No such file or directory):

    /mnt/g/qt/buildtrees/healpix/config-x64-linux-dbg-out.log
Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:388 (vcpkg_execute_required_process)
  ports/healpix/portfile.cmake:17 (vcpkg_configure_make)
```

Neumann-A committed:
remove trace from install

Neumann-A committed:
fix relative build path issues.

Neumann-A committed:
bump control of the 16 configure ports

Neumann-A committed:
never forget .... real linux is case sensitive ....

Neumann-A committed:
Revert "bump control of the 16 configure ports"

This reverts commit 40d6d81c01d2709c4acbdbec503b4787c3fd8282.

Neumann-A commented:
@ras0219-msft: Please tell us how to deal with the compiler flags in the different `configure_<scripts>` theoretically this should be all done from within the toolchain file!

Neumann-A committed:
Revert "Revert "bump control of the 16 configure ports""

This reverts commit 9c9851ddfcd01bfad604ab9c9aed7379238d619a.

Neumann-A committed:
bump control again for real ci rebuild

Neumann-A committed:
add copy_source

Neumann-A committed:
remove message

Neumann-A committed:
pass parameters to vcpkg_build_make

Neumann-A committed:
fix healpix build.

Neumann-A committed:
fix libmagic regression

Neumann-A committed:
fix libwandio regression

Neumann-A committed:
pfring changes regression fix? (cannot test in wsl)

Neumann-A committed:
ws change to retrigger CI

Voskrese commented:
> no package under MSVC is compiled, each one has an error in the package code
> 
> the edits themselves for the build from MAKE are working
> 
> plus the problem of finding libraries, in the absence of pkgconfig (*.pc)

Voskrese commented:
not work **MSVC** test `vcpkg install antioch:x64-windows`

```
find_program(GIT NAMES git git.cmd)
set(GIT_URL "https://github.com/libantioch/antioch")
set(GIT_REV master)
set(SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/${PORT})
if(NOT EXISTS "${SOURCE_PATH}/.git")
    message(STATUS "Cloning")
    vcpkg_execute_required_process(
      COMMAND ${GIT} clone --recurse-submodules -q --depth=20 --branch=${GIT_REV} ${GIT_URL} ${SOURCE_PATH}
#      COMMAND ${GIT} clone submodule sync --recursive -q --depth=20 --branch=${GIT_REV} ${GIT_URL} ${SOURCE_PATH}
      WORKING_DIRECTORY ${SOURCE_PATH}
      LOGNAME clone
    )
    message(STATUS "Fetching submodules")
    vcpkg_execute_required_process(
      COMMAND ${GIT} submodule update --init --recursive
      WORKING_DIRECTORY ${SOURCE_PATH}
      LOGNAME submodules
    )
endif()

vcpkg_configure_make(
    SOURCE_PATH ${SOURCE_PATH}
    COPY_SOURCE
    PRERUN_SHELL ${SOURCE_PATH}/bootstrap
    OPTIONS
      "--with-eigen-include='\${prefix}'/include"
    OPTIONS_DEBUG
      "--with-boost='\${prefix}'/debug/lib"
    OPTIONS_RELEASE
      "--with-boost='\${prefix}'/lib"
)

vcpkg_install_make()
```

Voskrese commented:
@dan-shaw 
merge, I see no reason for the delay, everything works, it is not possible to test ports without merging

Neumann-A commented:
no merge unless the libpq regression on linux is fixed. 
I don't quite understand why it is happening. I added COPY_SOURCE but it still does not work.....

Neumann-A committed:
fix libpq regression

Neumann-A committed:
fix libudns regression

Neumann-A committed:
add share/pkgconfig to PKG_CONFIG_PATH

Neumann-A committed:
cleanup of deprecated options

Neumann-A committed:
cleanup docs

JackBoosY commented on scripts/cmake/vcpkg_configure_make.cmake:
```diff
@@ -73,300 +53,453 @@
-    if(_csc_GENERATOR MATCHES "NMake")
-        message(FATAL_ERROR "Sorry, NMake does not supported currently.")
-        if (CMAKE_HOST_WIN32)
-            set(GENERATOR "nmake")
```
I had envisaged to provide the use of nmake to build ports in the future.

Neumann-A replied:
unless the makefile was generated with nmake in mind you will be greeted by syntax errors. As such errors where common and I completely removed it.

PhoebeHui commented on ports/farmhash/portfile.cmake:
```diff
@@ -1,4 +1,3 @@
-vcpkg_fail_port_install(ON_ARCH "arm" "arm64" ON_TARGET "UWP" "Windows")
```
@Neumann-A , thanks for the improvements!

Does it start to support the arch and targets? if so, these also need to be removed from the ./scripts/ci.baseline.txt file?

PhoebeHui commented on ports/farmhash/CONTROL:
```diff
@@ -1,4 +1,4 @@
 Source: farmhash
-Version: 1.1
+Version: 1.1-2
```
Could you update the Version to  '1.1-1'?  Also update the similar changes for the updated ports here.

PhoebeHui commented on ports/libosip2/portfile.cmake:
```diff
@@ -15,22 +15,19 @@ vcpkg_extract_source_archive_ex(
+        # file(COPY ${LIBOSIP2_BINARY} DESTINATION ${CURRENT_PACKAGES_DIR}/bin)
+    # endif()
+    # file(REMOVE ${LIBOSIP2_BINARY})
+# endforeach()
```
Could you please remove these code if they doesn't need any more?

Neumann-A commented on scripts/cmake/vcpkg_configure_make.cmake:
```diff
@@ -73,300 +53,453 @@
+    set(SRC_DIR "${_csc_SOURCE_PATH}/${_csc_PROJECT_SUBPATH}")
+
+    # Run autoconf if necessary
+    if(EXISTS "${SRC_DIR}/configure" AND NOT _csc_SKIP_CONFIGURE)
```
running autotools `configure.ac` is used to generate `configure` bash script. Unless you patch the `configure.ac` there is no need/reason to rerun autoconf. A valid use case to patch the configure.ac file is to patch AC_CHECK_LIB to AC_SEARCH_LIBS [(man)](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Libraries.html) for libraries like zlib (z, zlib zlibd) or bzip2 (bz2, bz2d).
Autoconf is automatically ignored if the is a `configure` bash script.

PhoebeHui reviewed (changes requested):
@Neumann-A, thanks as always!

Some thoughts:
1. The Version in CONTROL file need to be changed, the suffix of the version start from '-1'.
2. If the ports start to support arch or targets, then these port also need to remove from ./scripts/ci.baseline.txt file, the CI test system would help test these ports.
3. The features of the port doesn't test by CI testing system, so we need test it locally before remove them from ci.baseline.txt file. let us know If you need help.
4. If the some code doesn't need any more, remove it instead.
5. Since it contains large feature changes in this PR, so these changes need more experienced team members or contributors to review before merge.

Neumann-A committed:
Revert "bump control again for real ci rebuild"

This reverts commit b4bc18edc2376eba6c3aa3dda52bd82286f23616.

Neumann-A committed:
shared binaries on linux should stay in lib. Only dlls are mvoed to bin

Neumann-A commented:
>If the ports start to support arch or targets, then these port also need to remove from ./scripts/ci.baseline.txt file, the CI test system would help test these ports.

would have triggered an CI error if they do not fail (unless they are set to skip)

>The features of the port doesn't test by CI testing system, so we need test it locally before remove them from ci.baseline.txt file. 

Most ports which do not currently build on windows, probably build on windows by either using a different compiler (e.g. errors like __builtin bla bla not defined-> switch to gcc or clang-cl) or by using mingw (e.g. if posix like headers are missing). As such having a hard failure in the portfile is a no-go

> let us know If you need help.

feel free to change the PR as you like. I mean there is a reason I allow edits from maintainers. :)
I always wonder why you are so conservative about changing PRs. If it is something simple like updating the baseline or changing a control file i would expect more activity from your side instead of waiting until the author updates these.

dan-shaw committed:
Merge branch 'master' into improve_make_builds

Voskrese commented on scripts/cmake/vcpkg_configure_make.cmake:
```diff
@@ -73,300 +53,453 @@
+            set(ENV{PKG_CONFIG} "${PKGCONFIG} --define-variable=prefix=${_VCPKG_INSTALLED}/debug")
+            
+            set(dbg_command
+                ${base_cmd} -c "${CONFIGURE_ENV} ./${RELATIVE_BUILD_PATH}/configure ${BUILD_TARGET} ${HOST_TYPE}${_csc_OPTIONS} ${_csc_OPTIONS_DEBUG}")
```
```suggestion
                ${base_cmd} -c "${CONFIGURE_ENV} ./${SRC_DIR}/configure ${BUILD_TARGET} ${HOST_TYPE}${_csc_OPTIONS} ${_csc_OPTIONS_DEBUG}")

```
/.. **/./** ../ - error **perl** search path \\ and in other similar cases
SRC_DIR work path

Neumann-A replied:
Whats exactly the problem here or when exactly does the problem occur (so that i can check it)? I use `RELATIVE_BUILD_PATH` because it is certain to not have any spaces in it while `SRC_DIR` could possible have spaces screwing up the call.

Voskrese replied:
https://github.com/microsoft/vcpkg/pull/10402#issuecomment-599155720

see error in `config.log`

Voskrese replied:
this can be seen during any compilation, but the error will only be returned if additional utilities such as PERL are used

> это можно увидеть при любой компиляции, но ошибку выдаст только если используется дополнительные утилиты, типа PERL

Neumann-A replied:
@Voskrese: I don't know how to reproduce your error. And I tried with #10402 (comment) but ran into other issues with the mentioned portfiles. 
Please add the `<port>/CONTROL` and `<port>/portfile.cmake` you used so that I can have a look on the errors

Neumann-A committed:
move pkgconfig check and add message about missing system packages

Voskrese commented:
test
erorr

> Makefile:122: update target 'install' due to: all force
> /bin/sh E:/tools/vcpkg/buildtrees/gperf-tools/src/gperf-3-ccd5daa692/src/../build-aux/mkinstalldirs E:/tools/vcpkg/packages/gperf-tools_x64-windowsE:/tools/vcpkg/installed/x64-windows/debug/../tools/gperf-tools/debug/bin
> INSTALL_PROGRAM@ gperf.exe E:/tools/vcpkg/packages/gperf-tools_x64-windowsE:/tools/vcpkg/installed/x64-windows/debug/../tools/gperf-tools/debug/bin/gperf.exe
> make[1]: Leaving directory '/e/tools/vcpkg/buildtrees/gperf-tools/x64-windows-dbg/src'

`DESTDIR=${CURRENT_PACKAGES_DIR}` and `--sbindir=` conflict

JackBoosY commented:
/azp run

Neumann-A committed:
added autopoint as a build requirement

Neumann-A committed:
fix prerun shell working dir.

Voskrese commented:
@Neumann-A gperf-tools
only exe & doc [deps: texlive (perl)]

[gperf-tools.zip](https://github.com/microsoft/vcpkg/files/4387828/gperf-tools.zip)

Voskrese commented:
@Neumann-A error log

[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/4388852/install-x64-windows-dbg-out.log)

1. `../`

/bin/sh `.././../`src/gperf-3-ccd5daa692/src/../build-aux/mkinstalldirs

2. `/E` =  `DESTDIR=${_VCPKG_PACKAGE_PREFIX})` + `"--bindir=${prefix_var}/tools/${PORT}/bin"`

/E/tools/vcpkg/packages/gperf-tools_x64-`windows/E/tools`/vcpkg/installed/x64-windows/debug/../tools/gperf-tools/debug/bin

Voskrese commented on scripts/cmake/vcpkg_build_make.cmake:
```diff
@@ -159,27 +228,44 @@ function(vcpkg_build_make)
+        string(REGEX REPLACE "([a-zA-Z]):/" "/\\1/" _VCPKG_INSTALL_PREFIX "${CURRENT_INSTALLED_DIR}")
+        file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}_tmp")
+        file(RENAME "${CURRENT_PACKAGES_DIR}" "${CURRENT_PACKAGES_DIR}_tmp")
+        file(RENAME "${CURRENT_PACKAGES_DIR}_tmp${_VCPKG_INSTALL_PREFIX}/" "${CURRENT_PACKAGES_DIR}")
```
?
```suggestion
        file(RENAME "${CURRENT_PACKAGES_DIR}_tmp" "${CURRENT_PACKAGES_DIR}")

```

  file RENAME failed to rename
    E:/tools/vcpkg/packages/gperf-tools_x64-windows_tmp/E/tools/vcpkg/installed/x64-windows/
  to
    E:/tools/vcpkg/packages/gperf-tools_x64-windows
  because: No such file or directory

Neumann-A replied:
A few details here why it is how it is: 
- prefix must be pointed to the final install location which is `CURRENT_INSTALLED_DIR`
- first "installation" must be into `CURRENT_PACKAGES_DIR` due to vcpkg rules how it stages installs
- this is achieved by using [DESTDIR](https://www.gnu.org/prep/standards/html_node/DESTDIR.html) and setting it to `CURRENT_PACKAGES_DIR`
- thus make install goes into ${CURRENT_PACKAGES_DIR}/${CURRENT_INSTALLED_DIR} which is equivalent to DESTDIR/PREFIX
- since VCPKG does not support such a folder hierarchy `${CURRENT_PACKAGES_DIR}/${CURRENT_INSTALLED_DIR}/<installed_folders>` must be renamed to `${CURRENT_PACKAGES_DIR}/<installed_folders>`


>file RENAME failed to rename
E:/tools/vcpkg/packages/gperf-tools_x64-windows_tmp/E/tools/vcpkg/installed/x64-windows/

This probably means nothing has been installed and might indicate that a manual installation is needed by calling `vcpkg_build_make` instead of `vcpkg_install_make`. Another possibility is that the makefile is not following GNU coding standards and installs into some other folder without `DESTDIR` support and should be patched to follow coding standards.

Voskrese replied:
> DESTDIR=${_VCPKG_PACKAGE_PREFIX}) + "--bindir=${prefix_var}/tools/${PORT}/bin"
> =
> E:/tools/vcpkg/packages/gperf-tools_x64-windows_tmp/E/tools/vcpkg/installed/x64-windows/
> rename to
>  E:/tools/vcpkg/packages/gperf-tools_x64-windows


accessibly

Neumann-A committed:
a few fixes from the x windows pr
- add libpath setting
- fixed pkgconfig search on windows platforms
- fixed autopoint msys package on windows

Neumann-A committed:
fix release prefix

JackBoosY commented:
```
make: makefile: No such file or directory
make: *** No rule to make target 'makefile'.  Stop.
```
Some path are not set?

Neumann-A commented:
@JackBoosY: Sure you used the latest commit?
```
make: makefile: No such file or directory
make: *** No rule to make target 'makefile'.  Stop.
```
Indicates a lower/upper case error which i fixed in cb536ec. 
Bit annoying to change the scripts since CI does not rebuild the ports with those changes.

JackBoosY commented:
/azp run

Neumann-A commented:
@JackBoosY: Did not rerun. 
```
Triplet: x64-linux
Total elapsed time: 1.7 us
RESULTS
```

Neumann-A committed:
minimal cleanup and ws changes to trigger true CI rebuild

Neumann-A commented:
ah I forgot that changing libpq's portfile triggers a 7h linux rebuild due to Qt being rebuild.....

Neumann-A commented:
@JackBoosY. The OSX CI failed to publish the failure log

Neumann-A committed:
fix tcl by upgrading

Neumann-A committed:
remove unnecessary rename

Neumann-A committed:
fix pfring? can't test in WSL. Linux kernel headers required.

Neumann-A committed:
fix lowercase in patch

Neumann-A committed:
fix libwandio

Neumann-A committed:
remove x264 from fail list

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into improve_make_builds

JackBoosY commented:
I noticed that you commented `vcpkg_fail_port_install` in many places without modifying the baseline. Is this reasonable?

Neumann-A commented:
> I noticed that you commented vcpkg_fail_port_install in many places without modifying the baseline. Is this reasonable?

`vcpkg_fail_port_install` is too restrictive if used with WINDOWS as a target. Some of the ports fail because they can only be compiled with gcc or clang due to compiler specific syntax other ports might be able to compile with windows POSIX headers like unistd.h and others and could be compiled with MinGW/MSYS. 
Since 
```
elseif(VCPKG_CMAKE_SYSTEM_NAME STREQUAL "MinGW")
    set(VCPKG_TARGET_IS_WINDOWS 1)
    set(VCPKG_TARGET_IS_MINGW 1)
endif()
```
is defined we would block compiling with MinGW in these cases.

emptyVoid commented:
I would suggest merging debug and release configure paths in `vcpkg_configure_make.cmake` into a ```foreach(BUILDTYPE "debug" "release")``` in the same manner as in `vcpkg_build_make.cmake`, since the differences are minimal.

Neumann-A committed:
replace wrappers with wrappers from automake

BillyONeal commented on ports/libmagic/portfile.cmake:
```diff
@@ -1,4 +1,4 @@
-vcpkg_fail_port_install(MESSAGE "${PORT} currently only supports Linux and Mac platform" ON_TARGET "Windows")
+#vcpkg_fail_port_install(MESSAGE "${PORT} currently only supports Linux and Mac platform" ON_TARGET "Windows") 
```
Don't think we should leave commented out stuff in source control...

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into improve_make_builds

Neumann-A committed:
update make ports

Neumann-A committed:
remove unnecessary message

Neumann-A committed:
refactor vcpkg_configure_make

JackBoosY commented:
Pinging @Neumann-A  for response. Is work still being done for this PR?

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into improve_make_builds

Neumann-A committed:
[tcl] remove stray ?

Neumann-A committed:
[farmhash] fix build

Neumann-A committed:
[freexl] ws change to retrigger ci build

Neumann-A committed:
[x264] add pthread as a dependency

Neumann-A committed:
[vcpkg/scripts] add ignore flag correctly

Neumann-A committed:
[sdl] fix vcpkg_fixup_pkgconfig call

JackBoosY commented:
REGRESSION:
- freexl:x64-osx
- farmhash:x64-linux
- libudns:x64-linux

Neumann-A committed:
[farmhash/freexl] delete configure and recreate it.

Neumann-A committed:
[libudns] remove trailing ?

Neumann-A committed:
[freexl] use empty build target

Neumann-A committed:
[freexl] add system library; remove previous change

Voskrese commented:
@Neumann-A

had a good start, it would have been good to finish and combine

JackBoosY commented on ports/libwandio/portfile.cmake:
```diff
@@ -1,20 +1,18 @@
-
 vcpkg_install_make()
-
+#vcpkg_fixup_pkgconfig()?
```
I think we can remove the comment.

JackBoosY commented on ports/x264/portfile.cmake:
```diff
@@ -1,6 +1,6 @@
 set(X264_VERSION 157)
 
-vcpkg_fail_port_install(ON_TARGET "Linux" "OSX")
+#vcpkg_fail_port_install(ON_TARGET "Linux" "OSX") 
```
```suggestion
vcpkg_fail_port_install(ON_TARGET "OSX") 
```

Based on the baseline, I think x264 still does not support osx.

Neumann-A committed:
Update ports/x264/portfile.cmake

Co-authored-by: Jack·Boos·Yu <47264268+JackBoosY@users.noreply.github.com>

Neumann-A committed:
[libwandio] remove unncessary comment

JackBoosY commented:
Ping @strega-nil for review this PR.

strega-nil commented:
Alright, thanks so much @Neumann-A, this is incredible!

dg0yt commented:
This PR changed the pkg-config configuration from using `PKG_CONFIG_PATH` to `--define-variable=prefix=...`, and I wonder why this was done.

I'm asking because I see a compilation error for GDAL on Linux related to OpenEXR. OpenEXR seems to be installed in the Linux image (Azure pipelines), as it is found via pkg-config by GDAL, but not a dependency in vcpkg. There may be an insufficient configuration test on the gdal side (it tests linking, but not including headers). On the vcpkg side, IIUC the consequence of `--define-variable=prefix=...` is that .pc files are found in the system directories, but headers and libraries are redirected to the vcpkg installation. I.e. what is meant as `/usr/lib/x86_64-linux-gnu` by the system .pc file becomes `/home/vsts/work/1/s/installed/x64-linux/lib/x86_64-linux-gnu` when used in vcpkg. Is this intented? 

I suppose what is desired is that .pc files, includes and libraries are used from vcpkg only, similar to cross-compilation contexts. In my experience, this requires setting `PKG_CONFIG_PATH`, `PKG_CONFIG_LIBDIR` and `PKG_CONFIG_SYSROOT_DIR`, and this is often done in wrapper scripts. Cf. https://autotools.io/pkgconfig/cross-compiling.html

Neumann-A commented:
@dg0yt at the time vcpkg_configure_make was not really useable on any platform. Pc files were not corrected and the prefix was a hardcoded absolute path. The --define-variable was introduced before $(pcfiledir) was known about and integrated into vcpkg_fixup_pkgconfig. Before that the prefix variable was completly removed which made setting it a requirement. If it is still in vcpkg_configure_make and creating problems just remove it. It is no longer necessary

dg0yt commented:
@Neumann-A Thank you. I will try without. Looking forward to test with Android triplets soon.
